File: libs/aacl-go/aacl.go
package aacl

import (
	"crypto/ed25519"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"time"

	"github.com/aidenlippert/zerostate/libs/agentcard-go"
	"github.com/google/uuid"
)

// MessageType defines the type of AACL message
type MessageType string

const (
	MessageTypeRequest          MessageType = "Request"
	MessageTypeResponse         MessageType = "Response"
	MessageTypeQuery            MessageType = "Query"
	MessageTypeNotification     MessageType = "Notification"
	MessageTypeNegotiation      MessageType = "Negotiation"
	MessageTypeError            MessageType = "Error"
	MessageTypeAcknowledgment   MessageType = "Acknowledgment"
	MessageTypeWorkflowRequest  MessageType = "WorkflowRequest"
	MessageTypeWorkflowResponse MessageType = "WorkflowResponse"
	MessageTypeWorkflowStatus   MessageType = "WorkflowStatus"
	MessageTypeStreaming        MessageType = "Streaming"
)

// Intent describes what the user wants to accomplish
type Intent struct {
	Action               string                 `json:"action"`
	Goal                 string                 `json:"goal"`
	NaturalLanguage      *string                `json:"natural_language,omitempty"`
	Parsed               map[string]interface{} `json:"parsed,omitempty"`
	Confidence           *float64               `json:"confidence,omitempty"`
	CapabilitiesRequired []string               `json:"capabilities_required,omitempty"`
	Parameters           map[string]interface{} `json:"parameters"`
}

// IntentBuilder helps construct Intents
type IntentBuilder struct {
	action               string
	goal                 string
	naturalLanguage      *string
	parsed               map[string]interface{}
	confidence           *float64
	capabilitiesRequired []string
	parameters           map[string]interface{}
}

// NewIntent creates a new IntentBuilder
func NewIntent(action, goal string) *IntentBuilder {
	return &IntentBuilder{
		action:     action,
		goal:       goal,
		parameters: make(map[string]interface{}),
	}
}

// WithNaturalLanguage sets the natural language description
func (b *IntentBuilder) WithNaturalLanguage(nl string) *IntentBuilder {
	b.naturalLanguage = &nl
	return b
}

// WithParsed sets parsed structured data
func (b *IntentBuilder) WithParsed(parsed map[string]interface{}) *IntentBuilder {
	b.parsed = parsed
	return b
}

// WithConfidence sets the confidence score
func (b *IntentBuilder) WithConfidence(conf float64) *IntentBuilder {
	b.confidence = &conf
	return b
}

// RequiresCapability adds a required capability
func (b *IntentBuilder) RequiresCapability(cap string) *IntentBuilder {
	b.capabilitiesRequired = append(b.capabilitiesRequired, cap)
	return b
}

// WithParameter adds a parameter
func (b *IntentBuilder) WithParameter(key string, value interface{}) *IntentBuilder {
	b.parameters[key] = value
	return b
}

// Build constructs the Intent
func (b *IntentBuilder) Build() Intent {
	return Intent{
		Action:               b.action,
		Goal:                 b.goal,
		NaturalLanguage:      b.naturalLanguage,
		Parsed:               b.parsed,
		Confidence:           b.confidence,
		CapabilitiesRequired: b.capabilitiesRequired,
		Parameters:           b.parameters,
	}
}

// ExecutionMetadata tracks execution details
type ExecutionMetadata struct {
	DurationMs      uint64  `json:"duration_ms"`
	GasUsed         uint64  `json:"gas_used"`
	CostUainur      uint64  `json:"cost_uainur"`
	AgentVersion    string  `json:"agent_version"`
	AgentTrustScore float64 `json:"agent_trust_score"`
	ExecutionNodeID string  `json:"execution_node_id"`
	RetryCount      uint32  `json:"retry_count,omitempty"`
}

// ResponseResult represents the result of an execution
type ResponseResult struct {
	Value      interface{} `json:"value"`
	Type       string      `json:"type,omitempty"`
	Confidence *float64    `json:"confidence,omitempty"`
}

// ResponsePayload is the payload for Response messages
type ResponsePayload struct {
	Status            string             `json:"status"`
	Result            *ResponseResult    `json:"result,omitempty"`
	Error             *ErrorInfo         `json:"error,omitempty"`
	ExecutionMetadata *ExecutionMetadata `json:"execution_metadata,omitempty"`
}

// ErrorInfo describes an error
type ErrorInfo struct {
	Code                string                 `json:"code"`
	Message             string                 `json:"message"`
	Details             map[string]interface{} `json:"details,omitempty"`
	Recoverable         bool                   `json:"recoverable"`
	RecoverySuggestions []string               `json:"recovery_suggestions,omitempty"`
}

// ConversationContext maintains conversation state
type ConversationContext struct {
	ConversationID   string                 `json:"conversation_id"`
	PreviousMessages []string               `json:"previous_messages,omitempty"`
	SharedState      map[string]interface{} `json:"shared_state,omitempty"`
}

// WorkflowStep represents a step in a workflow
type WorkflowStep struct {
	StepID      string                 `json:"step_id"`
	AgentDID    agentcard.DID          `json:"agent_did"`
	Intent      Intent                 `json:"intent"`
	DependsOn   []string               `json:"depends_on,omitempty"`
	Timeout     *uint64                `json:"timeout,omitempty"`
	RetryPolicy map[string]interface{} `json:"retry_policy,omitempty"`
}

// Workflow orchestrates multi-agent tasks
type Workflow struct {
	WorkflowID   string                 `json:"workflow_id"`
	Goal         string                 `json:"goal"`
	Steps        []WorkflowStep         `json:"steps"`
	Dependencies map[string][]string    `json:"dependencies"`
	Metadata     map[string]interface{} `json:"metadata,omitempty"`
}

// AgentMatch represents a discovered agent
type AgentMatch struct {
	AgentDID     agentcard.DID `json:"agent_did"`
	TrustScore   float64       `json:"trust_score"`
	Price        uint64        `json:"price"`
	Latency      uint64        `json:"latency_ms"`
	Capabilities []string      `json:"capabilities"`
}

// CapabilityQuery queries for agents with specific capabilities
type CapabilityQuery struct {
	RequiredCapabilities []string `json:"required_capabilities"`
	MinTrustScore        *float64 `json:"min_trust_score,omitempty"`
	MaxPrice             *uint64  `json:"max_price,omitempty"`
	MaxLatency           *uint64  `json:"max_latency_ms,omitempty"`
	PreferredRegions     []string `json:"preferred_regions,omitempty"`
}

// AACLMessage is the core message structure
type AACLMessage struct {
	Context             string               `json:"@context"`
	Type                string               `json:"@type"`
	ID                  string               `json:"id"`
	From                agentcard.DID        `json:"from"`
	To                  agentcard.DID        `json:"to"`
	Timestamp           time.Time            `json:"timestamp"`
	Intent              *Intent              `json:"intent,omitempty"`
	Payload             interface{}          `json:"payload"`
	ConversationContext *ConversationContext `json:"conversation_context,omitempty"`
	Signature           *string              `json:"signature,omitempty"`
}

// MessageBuilder helps construct AACLMessages
type MessageBuilder struct {
	msgType             MessageType
	from                agentcard.DID
	to                  agentcard.DID
	intent              *Intent
	payload             interface{}
	conversationContext *ConversationContext
}

// NewMessage creates a new MessageBuilder
func NewMessage(msgType MessageType, from, to agentcard.DID) *MessageBuilder {
	return &MessageBuilder{
		msgType: msgType,
		from:    from,
		to:      to,
	}
}

// WithIntent sets the intent
func (b *MessageBuilder) WithIntent(intent Intent) *MessageBuilder {
	b.intent = &intent
	return b
}

// WithPayload sets the payload
func (b *MessageBuilder) WithPayload(payload interface{}) *MessageBuilder {
	b.payload = payload
	return b
}

// WithConversationContext sets conversation context
func (b *MessageBuilder) WithConversationContext(ctx ConversationContext) *MessageBuilder {
	b.conversationContext = &ctx
	return b
}

// Build constructs the AACLMessage
func (b *MessageBuilder) Build() (*AACLMessage, error) {
	if b.from == "" {
		return nil, fmt.Errorf("from DID is required")
	}
	if b.to == "" {
		return nil, fmt.Errorf("to DID is required")
	}

	return &AACLMessage{
		Context:             "https://ainur.network/contexts/aacl/v1",
		Type:                string(b.msgType),
		ID:                  fmt.Sprintf("urn:uuid:%s", uuid.New().String()),
		From:                b.from,
		To:                  b.to,
		Timestamp:           time.Now(),
		Intent:              b.intent,
		Payload:             b.payload,
		ConversationContext: b.conversationContext,
		Signature:           nil,
	}, nil
}

// RequestMessage creates a Request message
func RequestMessage(from, to agentcard.DID, intent Intent) (*AACLMessage, error) {
	return NewMessage(MessageTypeRequest, from, to).
		WithIntent(intent).
		Build()
}

// ResponseMessage creates a Response message
func ResponseMessage(from, to agentcard.DID, payload ResponsePayload) (*AACLMessage, error) {
	return NewMessage(MessageTypeResponse, from, to).
		WithPayload(payload).
		Build()
}

// ErrorMessage creates an Error message
func ErrorMessage(from, to agentcard.DID, errorInfo ErrorInfo) (*AACLMessage, error) {
	return NewMessage(MessageTypeError, from, to).
		WithPayload(errorInfo).
		Build()
}

// WorkflowRequestMessage creates a WorkflowRequest message
func WorkflowRequestMessage(from, to agentcard.DID, workflow Workflow) (*AACLMessage, error) {
	return NewMessage(MessageTypeWorkflowRequest, from, to).
		WithPayload(workflow).
		Build()
}

// Sign signs an AACL message with an Ed25519 private key
func (m *AACLMessage) Sign(privateKey ed25519.PrivateKey) error {
	// Remove existing signature
	m.Signature = nil

	// Create canonical JSON
	msgJSON, err := json.Marshal(m)
	if err != nil {
		return fmt.Errorf("failed to marshal message: %w", err)
	}

	// Sign the JSON
	signature := ed25519.Sign(privateKey, msgJSON)

	// Encode signature as base64
	signatureStr := base64.StdEncoding.EncodeToString(signature)
	m.Signature = &signatureStr

	return nil
}

// Verify verifies the message signature
func (m *AACLMessage) Verify(publicKey ed25519.PublicKey) (bool, error) {
	if m.Signature == nil {
		return false, fmt.Errorf("message has no signature")
	}

	// Decode signature
	signature, err := base64.StdEncoding.DecodeString(*m.Signature)
	if err != nil {
		return false, fmt.Errorf("failed to decode signature: %w", err)
	}

	// Remove signature for verification
	sigBackup := m.Signature
	m.Signature = nil
	msgJSON, err := json.Marshal(m)
	m.Signature = sigBackup
	if err != nil {
		return false, fmt.Errorf("failed to marshal message: %w", err)
	}

	// Verify signature
	return ed25519.Verify(publicKey, msgJSON, signature), nil
}

// ToJSON converts the message to JSON
func (m *AACLMessage) ToJSON() ([]byte, error) {
	return json.MarshalIndent(m, "", "  ")
}

// FromJSON parses an AACLMessage from JSON
func FromJSON(data []byte) (*AACLMessage, error) {
	var msg AACLMessage
	if err := json.Unmarshal(data, &msg); err != nil {
		return nil, err
	}
	return &msg, nil
}

// SuccessResponse creates a successful ResponsePayload
func SuccessResponse(result interface{}, metadata *ExecutionMetadata) ResponsePayload {
	return ResponsePayload{
		Status: "success",
		Result: &ResponseResult{
			Value: result,
		},
		ExecutionMetadata: metadata,
	}
}

// ErrorResponse creates an error ResponsePayload
func ErrorResponse(errorInfo ErrorInfo) ResponsePayload {
	return ResponsePayload{
		Status: "error",
		Error:  &errorInfo,
	}
}

// NewConversation creates a new ConversationContext
func NewConversation() ConversationContext {
	return ConversationContext{
		ConversationID:   uuid.New().String(),
		PreviousMessages: []string{},
		SharedState:      make(map[string]interface{}),
	}
}

// AddMessage adds a message ID to the conversation history
func (c *ConversationContext) AddMessage(messageID string) {
	c.PreviousMessages = append(c.PreviousMessages, messageID)
}

// SetState sets a value in the shared state
func (c *ConversationContext) SetState(key string, value interface{}) {
	c.SharedState[key] = value
}

// GetState retrieves a value from shared state
func (c *ConversationContext) GetState(key string) (interface{}, bool) {
	val, ok := c.SharedState[key]
	return val, ok
}

File: examples/python-sdk/aacl.py
"""
AACL-v1 Python SDK

Simple Python implementation for creating and working with AACL messages
(Ainur Agent Communication Language).
"""

import json
from datetime import datetime
from typing import Dict, List, Optional, Any
from dataclasses import dataclass, field
from enum import Enum
import uuid


class MessageType(Enum):
    """AACL message types"""
    REQUEST = "Request"
    RESPONSE = "Response"
    QUERY = "Query"
    NOTIFICATION = "Notification"
    NEGOTIATION = "Negotiation"
    ERROR = "Error"
    ACKNOWLEDGMENT = "Acknowledgment"
    WORKFLOW_REQUEST = "WorkflowRequest"
    WORKFLOW_RESPONSE = "WorkflowResponse"
    WORKFLOW_STATUS = "WorkflowStatus"
    STREAMING = "Streaming"


@dataclass
class Intent:
    """User intent description"""
    action: str
    goal: str
    natural_language: Optional[str] = None
    parsed: Optional[Dict[str, Any]] = None
    confidence: Optional[float] = None
    capabilities_required: List[str] = field(default_factory=list)
    parameters: Dict[str, Any] = field(default_factory=dict)
    
    def to_dict(self) -> Dict[str, Any]:
        """Convert to dictionary"""
        data = {
            "action": self.action,
            "goal": self.goal,
            "parameters": self.parameters
        }
        if self.natural_language:
            data["natural_language"] = self.natural_language
        if self.parsed:
            data["parsed"] = self.parsed
        if self.confidence is not None:
            data["confidence"] = self.confidence
        if self.capabilities_required:
            data["capabilities_required"] = self.capabilities_required
        return data


@dataclass
class ExecutionMetadata:
    """Execution metadata"""
    duration_ms: int
    gas_used: int
    cost_uainur: int
    agent_version: str
    agent_trust_score: float
    execution_node_id: str
    retry_count: int = 0
    
    def to_dict(self) -> Dict[str, Any]:
        """Convert to dictionary"""
        return {
            "duration_ms": self.duration_ms,
            "gas_used": self.gas_used,
            "cost_uainur": self.cost_uainur,
            "agent_version": self.agent_version,
            "agent_trust_score": self.agent_trust_score,
            "execution_node_id": self.execution_node_id,
            "retry_count": self.retry_count
        }


@dataclass
class ResponseResult:
    """Response result"""
    value: Any
    result_type: Optional[str] = None
    confidence: Optional[float] = None
    
    def to_dict(self) -> Dict[str, Any]:
        """Convert to dictionary"""
        data = {"value": self.value}
        if self.result_type:
            data["type"] = self.result_type
        if self.confidence is not None:
            data["confidence"] = self.confidence
        return data


@dataclass
class ResponsePayload:
    """Response message payload"""
    status: str
    result: Optional[ResponseResult] = None
    error: Optional['ErrorInfo'] = None
    execution_metadata: Optional[ExecutionMetadata] = None
    
    def to_dict(self) -> Dict[str, Any]:
        """Convert to dictionary"""
        data = {"status": self.status}
        if self.result:
            data["result"] = self.result.to_dict()
        if self.error:
            data["error"] = self.error.to_dict()
        if self.execution_metadata:
            data["execution_metadata"] = self.execution_metadata.to_dict()
        return data


@dataclass
class ErrorInfo:
    """Error information"""
    code: str
    message: str
    details: Optional[Dict[str, Any]] = None
    recoverable: bool = False
    recovery_suggestions: List[str] = field(default_factory=list)
    
    def to_dict(self) -> Dict[str, Any]:
        """Convert to dictionary"""
        data = {
            "code": self.code,
            "message": self.message,
            "recoverable": self.recoverable
        }
        if self.details:
            data["details"] = self.details
        if self.recovery_suggestions:
            data["recovery_suggestions"] = self.recovery_suggestions
        return data


@dataclass
class ConversationContext:
    """Conversation context"""
    conversation_id: str
    previous_messages: List[str] = field(default_factory=list)
    shared_state: Dict[str, Any] = field(default_factory=dict)
    
    def add_message(self, message_id: str):
        """Add a message to history"""
        self.previous_messages.append(message_id)
    
    def set_state(self, key: str, value: Any):
        """Set shared state"""
        self.shared_state[key] = value
    
    def get_state(self, key: str) -> Optional[Any]:
        """Get shared state"""
        return self.shared_state.get(key)
    
    def to_dict(self) -> Dict[str, Any]:
        """Convert to dictionary"""
        return {
            "conversation_id": self.conversation_id,
            "previous_messages": self.previous_messages,
            "shared_state": self.shared_state
        }


@dataclass
class WorkflowStep:
    """Workflow step"""
    step_id: str
    agent_did: str
    intent: Intent
    depends_on: List[str] = field(default_factory=list)
    timeout: Optional[int] = None
    retry_policy: Optional[Dict[str, Any]] = None
    
    def to_dict(self) -> Dict[str, Any]:
        """Convert to dictionary"""
        data = {
            "step_id": self.step_id,
            "agent_did": self.agent_did,
            "intent": self.intent.to_dict()
        }
        if self.depends_on:
            data["depends_on"] = self.depends_on
        if self.timeout:
            data["timeout"] = self.timeout
        if self.retry_policy:
            data["retry_policy"] = self.retry_policy
        return data


@dataclass
class Workflow:
    """Multi-agent workflow"""
    workflow_id: str
    goal: str
    steps: List[WorkflowStep]
    dependencies: Dict[str, List[str]] = field(default_factory=dict)
    metadata: Optional[Dict[str, Any]] = None
    
    def to_dict(self) -> Dict[str, Any]:
        """Convert to dictionary"""
        data = {
            "workflow_id": self.workflow_id,
            "goal": self.goal,
            "steps": [step.to_dict() for step in self.steps],
            "dependencies": self.dependencies
        }
        if self.metadata:
            data["metadata"] = self.metadata
        return data


@dataclass
class AACLMessage:
    """AACL message"""
    message_type: MessageType
    from_did: str
    to_did: str
    message_id: str
    timestamp: str
    intent: Optional[Intent] = None
    payload: Optional[Any] = None
    conversation_context: Optional[ConversationContext] = None
    signature: Optional[str] = None
    
    def to_dict(self) -> Dict[str, Any]:
        """Convert to dictionary with JSON-LD structure"""
        data = {
            "@context": "https://ainur.network/contexts/aacl/v1",
            "@type": self.message_type.value,
            "id": self.message_id,
            "from": self.from_did,
            "to": self.to_did,
            "timestamp": self.timestamp
        }
        
        if self.intent:
            data["intent"] = self.intent.to_dict()
        
        if self.payload:
            if hasattr(self.payload, 'to_dict'):
                data["payload"] = self.payload.to_dict()
            else:
                data["payload"] = self.payload
        
        if self.conversation_context:
            data["conversation_context"] = self.conversation_context.to_dict()
        
        if self.signature:
            data["signature"] = self.signature
        
        return data
    
    def to_json(self, indent: int = 2) -> str:
        """Convert to JSON string"""
        return json.dumps(self.to_dict(), indent=indent)
    
    @classmethod
    def from_dict(cls, data: Dict[str, Any]) -> 'AACLMessage':
        """Create from dictionary"""
        message_type = MessageType(data["@type"])
        
        intent = None
        if "intent" in data:
            intent_data = data["intent"]
            intent = Intent(
                action=intent_data["action"],
                goal=intent_data["goal"],
                natural_language=intent_data.get("natural_language"),
                parsed=intent_data.get("parsed"),
                confidence=intent_data.get("confidence"),
                capabilities_required=intent_data.get("capabilities_required", []),
                parameters=intent_data.get("parameters", {})
            )
        
        context = None
        if "conversation_context" in data:
            ctx_data = data["conversation_context"]
            context = ConversationContext(
                conversation_id=ctx_data["conversation_id"],
                previous_messages=ctx_data.get("previous_messages", []),
                shared_state=ctx_data.get("shared_state", {})
            )
        
        return cls(
            message_type=message_type,
            from_did=data["from"],
            to_did=data["to"],
            message_id=data["id"],
            timestamp=data["timestamp"],
            intent=intent,
            payload=data.get("payload"),
            conversation_context=context,
            signature=data.get("signature")
        )


class IntentBuilder:
    """Builder for creating Intents"""
    
    def __init__(self, action: str, goal: str):
        self.action = action
        self.goal = goal
        self.natural_language: Optional[str] = None
        self.parsed: Optional[Dict[str, Any]] = None
        self.confidence: Optional[float] = None
        self.capabilities_required: List[str] = []
        self.parameters: Dict[str, Any] = {}
    
    def with_natural_language(self, nl: str) -> 'IntentBuilder':
        """Set natural language description"""
        self.natural_language = nl
        return self
    
    def with_parsed(self, parsed: Dict[str, Any]) -> 'IntentBuilder':
        """Set parsed data"""
        self.parsed = parsed
        return self
    
    def with_confidence(self, confidence: float) -> 'IntentBuilder':
        """Set confidence score"""
        self.confidence = confidence
        return self
    
    def requires_capability(self, capability: str) -> 'IntentBuilder':
        """Add required capability"""
        self.capabilities_required.append(capability)
        return self
    
    def with_parameter(self, key: str, value: Any) -> 'IntentBuilder':
        """Add parameter"""
        self.parameters[key] = value
        return self
    
    def build(self) -> Intent:
        """Build the intent"""
        return Intent(
            action=self.action,
            goal=self.goal,
            natural_language=self.natural_language,
            parsed=self.parsed,
            confidence=self.confidence,
            capabilities_required=self.capabilities_required,
            parameters=self.parameters
        )


class MessageBuilder:
    """Builder for creating AACL messages"""
    
    def __init__(self, message_type: MessageType, from_did: str, to_did: str):
        self.message_type = message_type
        self.from_did = from_did
        self.to_did = to_did
        self.intent: Optional[Intent] = None
        self.payload: Optional[Any] = None
        self.conversation_context: Optional[ConversationContext] = None
    
    def with_intent(self, intent: Intent) -> 'MessageBuilder':
        """Set intent"""
        self.intent = intent
        return self
    
    def with_payload(self, payload: Any) -> 'MessageBuilder':
        """Set payload"""
        self.payload = payload
        return self
    
    def with_conversation_context(self, context: ConversationContext) -> 'MessageBuilder':
        """Set conversation context"""
        self.conversation_context = context
        return self
    
    def build(self) -> AACLMessage:
        """Build the message"""
        return AACLMessage(
            message_type=self.message_type,
            from_did=self.from_did,
            to_did=self.to_did,
            message_id=f"urn:uuid:{uuid.uuid4()}",
            timestamp=datetime.utcnow().isoformat() + "Z",
            intent=self.intent,
            payload=self.payload,
            conversation_context=self.conversation_context
        )


def create_request(from_did: str, to_did: str, intent: Intent) -> AACLMessage:
    """Create a Request message"""
    return MessageBuilder(MessageType.REQUEST, from_did, to_did).with_intent(intent).build()


def create_response(from_did: str, to_did: str, payload: ResponsePayload) -> AACLMessage:
    """Create a Response message"""
    return MessageBuilder(MessageType.RESPONSE, from_did, to_did).with_payload(payload).build()


def create_error(from_did: str, to_did: str, error_info: ErrorInfo) -> AACLMessage:
    """Create an Error message"""
    return MessageBuilder(MessageType.ERROR, from_did, to_did).with_payload(error_info).build()


def create_workflow_request(from_did: str, to_did: str, workflow: Workflow) -> AACLMessage:
    """Create a WorkflowRequest message"""
    return MessageBuilder(MessageType.WORKFLOW_REQUEST, from_did, to_did).with_payload(workflow).build()


def success_response(result: Any, metadata: Optional[ExecutionMetadata] = None) -> ResponsePayload:
    """Create a success response payload"""
    return ResponsePayload(
        status="success",
        result=ResponseResult(value=result),
        execution_metadata=metadata
    )


def error_response(error_info: ErrorInfo) -> ResponsePayload:
    """Create an error response payload"""
    return ResponsePayload(status="error", error=error_info)


def new_conversation() -> ConversationContext:
    """Create a new conversation context"""
    return ConversationContext(conversation_id=str(uuid.uuid4()))
